drivers/dma/dma_arc_hs.c

```diff
@@ -0,0 +1,1011 @@
+#define DMA_AUX_BASE     (0xd00)
+#define DMA_C_CTRL_AUX   (0xd00 + 0x0)
+#define DMA_C_CHAN_AUX   (0xd00 + 0x1)
+#define DMA_C_SRC_AUX    (0xd00 + 0x2)
+#define DMA_C_SRC_HI_AUX (0xd00 + 0x3)
+#define DMA_C_DST_AUX    (0xd00 + 0x4)
+#define DMA_C_DST_HI_AUX (0xd00 + 0x5)
+#define DMA_C_ATTR_AUX   (0xd00 + 0x6)
+#define DMA_C_LEN_AUX    (0xd00 + 0x7)
+#define DMA_C_HANDLE_AUX (0xd00 + 0x8)
+#define DMA_C_STAT_AUX   (0xd00 + 0xc)
+
+#define DMA_S_CTRL_AUX      (0xd00 + 0x10)
+#define DMA_S_BASEC_AUX(ch) (0xd00 + 0x83 + ((ch) * 8))
+#define DMA_S_LASTC_AUX(ch) (0xd00 + 0x84 + ((ch) * 8))
+#define DMA_S_STATC_AUX(ch) (0xd00 + 0x86 + ((ch) * 8))
+#define DMA_S_DONESTATD_AUX(d)                                                                     \
+	(0xd00 + 0x20 + (d)) /* Descriptor selection. Each D stores descriptors d*32 +: 32 */
+#define DMA_S_DONESTATD_CLR_AUX(d) (0xd00 + 0x40 + (d))
+
+/* ARC DMA Attribute Flags */
+#define ARC_DMA_NP_ATTR       (1 << 3) /* Enable non posted writes */
+#define ARC_DMA_SET_DONE_ATTR (1 << 0) /* Set done without triggering interrupt */
+
+/* ARC Auxiliary Register Access Functions */
+static inline void arc_write_aux(uint32_t reg, uint32_t val)
+{
+	__asm__ volatile("sr %0, [%1]" : : "r"(val), "r"(reg) : "memory");
+}
+
+static inline uint32_t arc_read_aux(uint32_t reg)
+{
+	uint32_t val;
+
+	__asm__ volatile("lr %0, [%1]" : "=r"(val) : "r"(reg) : "memory");
+	return val;
+}
+
+/* Low-level ARC DMA Functions */
+static inline void dma_arc_hs_config_hw(void)
+{
+	uint32_t reg = 0;
+
+	reg = (0xf << 4);                   /* Set LBU read transaction limit to max */
+	reg |= (0x4 << 8);                  /* Set max burst length to 16 (max supported) */
+	arc_write_aux(DMA_S_CTRL_AUX, reg); /* Apply settings above */
+}
+
+static inline void dma_arc_hs_init_channel_hw(uint32_t dma_ch, uint32_t base, uint32_t last)
+{
+	arc_write_aux(DMA_S_BASEC_AUX(dma_ch), base);
+	arc_write_aux(DMA_S_LASTC_AUX(dma_ch), last);
+	arc_write_aux(DMA_S_STATC_AUX(dma_ch), 0x1); /* Enable dma_ch */
+}
+
+static inline void dma_arc_hs_start_hw(uint32_t dma_ch, const void *p_src, void *p_dst,
+				       uint32_t len, uint32_t attr)
+{
+	arc_write_aux(DMA_C_CHAN_AUX, dma_ch);
+	arc_write_aux(DMA_C_SRC_AUX, (uint32_t)p_src);
+	arc_write_aux(DMA_C_DST_AUX, (uint32_t)p_dst);
+	arc_write_aux(DMA_C_ATTR_AUX, attr);
+	arc_write_aux(DMA_C_LEN_AUX, len);
+}
+
+/* Queue a transfer on the currently selected channel (for multi-block) */
+static inline void dma_arc_hs_next_hw(const void *p_src, void *p_dst, uint32_t len, uint32_t attr)
+{
+	/* Don't write DMA_C_CHAN_AUX - use currently selected channel */
+	arc_write_aux(DMA_C_SRC_AUX, (uint32_t)p_src);
+	arc_write_aux(DMA_C_DST_AUX, (uint32_t)p_dst);
+	arc_write_aux(DMA_C_ATTR_AUX, attr);
+	arc_write_aux(DMA_C_LEN_AUX, len);
+}
+
+static inline uint32_t dma_arc_hs_get_handle_hw(void)
+{
+	return arc_read_aux(DMA_C_HANDLE_AUX);
+}
+
+static inline uint32_t dma_arc_hs_poll_busy_hw(void)
```

Comment:
Can remove the `inline` modifier.
```suggestion
static uint32_t dma_arc_hs_poll_busy_hw(void)
```